File: android/src/main/kotlin/network/xyo/ble/xyo_ble/GattDefinedOperationRequest.kt
package network.xyo.ble.xyo_ble

import network.xyo.ble.devices.xy.XYFinderBluetoothDevice
import network.xyo.ble.flutter.protobuf.Gatt
import network.xyo.ble.generic.devices.XYBluetoothDevice
import network.xyo.ble.generic.gatt.peripheral.XYBluetoothResult
import network.xyo.modbluetoothkotlin.client.XyoSentinelX

class GattDefinedOperationHandler {

    companion object {
        // Run the operations
        @kotlin.ExperimentalUnsignedTypes
        suspend fun process(device: XYBluetoothDevice, operation: Gatt.DefinedOperation): XYBluetoothResult<Any>? {
                val finder = device as? XYFinderBluetoothDevice
                val sentinelX = device as? XyoSentinelX
                var result: XYBluetoothResult<Any>? = null
                device.connection {
                    when (operation) {
                        Gatt.DefinedOperation.SONG -> {
                            finder?.unlock()
                            finder?.find()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.STOP_SONG -> {
                            finder?.unlock()
                            finder?.stopFind()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.STAY_AWAKE -> {
                            finder?.unlock()
                            finder?.stayAwake()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.GO_TO_SLEEP -> {
                            finder?.unlock()
                            finder?.fallAsleep()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.LOCK -> {
                            finder?.lock()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.UNLOCK -> {
                            finder?.lock()?.let {
                                result = XYBluetoothResult(it.value, it.error)
                            }
                        }
                        Gatt.DefinedOperation.PUBLIC_KEY -> {
                            val key = sentinelX?.getPublicKey()
                            val value = key?.value
                            if (value != null) {
                                result = XYBluetoothResult(value.toBase58String())
                            } else {
                                result = XYBluetoothResult(false)
                            }
                        }
                        else -> {
                            result = XYBluetoothResult(false)
                        }
                    }
                    return@connection XYBluetoothResult(true)
                }
                return result
        }
    }
}

File: ios/Classes/GattDefinedOperationRequest.swift
import Flutter
import XyBleSdk
import CoreBluetooth
import SwiftProtobuf
import sdk_xyobleinterface_swift

protocol GattDefinedOperation {
    static func process(for device: XYBluetoothDevice, operation: DefinedOperation) -> XYBluetoothResult
}

struct GattDefinedOperationHandler: GattDefinedOperation {

    // Run the operations
    static func process(for device: XYBluetoothDevice, operation: DefinedOperation) -> XYBluetoothResult {
        if let xyDevice = self.deviceAsFinder(device) {
            switch operation {
            case .song:
                _ = xyDevice.unlock()
                return xyDevice.find(.findIt)
            case .stopSong:
                _ = xyDevice.unlock()
                return xyDevice.find(.off)
            case .stayAwake:
                _ = xyDevice.unlock()
                return xyDevice.stayAwake()
            case .goToSleep:
                _ = xyDevice.unlock
                return xyDevice.fallAsleep()
            case .lock:
                _ = xyDevice.unlock()
                return xyDevice.lock()
            case .unlock:
                return xyDevice.unlock()
            default:
                return XYBluetoothResult(error: .serviceNotFound)
            }
        } else if let xySentinelX = self.deviceAsSentinelX(device) {
            switch operation {
            case .publicKey:
                guard let key = xySentinelX.getPublicKey() else {
                    return XYBluetoothResult(error: .dataNotPresent)
                }
                // Convert from Base58 and then re-encode to utf8 for Dart
                return XYBluetoothResult(data: Data(bytes: Array(key.toBase58String().utf8)))
            default:
                return XYBluetoothResult(error: .serviceNotFound)
            }
        }


        return XYBluetoothResult(error: .actionNotSupported)
    }

    // Try to cast the device to an xy finder device
    private static func deviceAsFinder(_ device: XYBluetoothDevice) -> XYFinderDeviceBase? {
        if device.family.uuid == XY4BluetoothDevice.family.uuid {
            return device as? XY4BluetoothDevice
        }

        return nil
    }

    // Try to cast the device to a sentinel X device
    private static func deviceAsSentinelX(_ device: XYBluetoothDevice) -> XyoSentinelXDevice? {
        if device.family.uuid == XyoBluetoothDevice.family.uuid {
            return device as? XyoSentinelXDevice
        }

        return nil
    }

}
